### PR TITLE
Fix variational models for numpy backends

### PR DESCRIPTION
**Pull request by stavros11**
Variational model (VQE and QAOA) optimizations will fail if a numpy backend is used because there are some `.numpy()` calls in the loss function which are only supported by Tensorflow. This PR fixes this issue and extends the variational model optimization tests to numpy backends.

### CODE REVIEW

**Review by scarrazza — COMMENTED** [review]
The changes look ok, but from the tests failure, looks like there is something wrong when changing `K.op`.

**Review by scarrazza — APPROVED** [review]
Thanks for the fixes, looks good.
